oblast/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+pub struct Fp12 {
+    element: blst_fp12,
+}
+
+impl std::ops::Mul for Fp12 {
+    type Output = Fp12;
+
+    fn mul(mut self, rhs: Fp12) -> Fp12 {
+        unsafe {
+            blst::blst_fp12_mul(&mut self.element, &self.element, &rhs.element);
+        }
+        self
+    }
+}
+
+impl Fp12 {
+    pub fn from_raw(element: blst_fp12) -> Self {
+        Self { element }
+    }
+
+    pub fn final_exp(mut self) -> Self {
+        unsafe {
+            blst::blst_final_exp(&mut self.element, &self.element);
+        }
+        self
+    }
+
+    pub fn is_one(&self) -> bool {
+        unsafe { blst::blst_fp12_is_one(&self.element) }
+    }
+}
+
+/// Scalar for multiplying curve points.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct Scalar {
+    value: blst_scalar,
+}
+
+impl From<Fr> for Scalar {
+    fn from(x: Fr) -> Self {
+        let mut scalar = Self::default();
+        unsafe {
+            blst::blst_scalar_from_fr(&mut scalar.value, &x.element);
+        }
+        scalar
+    }
+}
+
+macro_rules! define_curve_struct {
+    ($struct_name:ident, $blst_name:ident, $group_name:ident, $compressed_bytes:expr) => {
+        paste! {
+            #[doc = "Point on the curve sub-group " $group_name "."]
+            #[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+            pub struct $struct_name {
+                point: blst::[<blst_ $blst_name>],
+            }
+        }
+
+        paste! {
+            #[doc = "Affine encoding of a point on " $group_name "."]
+            #[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+            pub struct [<$struct_name Affine>] {
+                point: blst::[<blst_ $blst_name _affine>],
+            }
+
+            impl From<&$struct_name> for [<$struct_name Affine>] {
+                fn from(point: &$struct_name) -> Self {
+                    let mut affine = Self::default();
+                    unsafe {
+                        blst::[<blst_ $blst_name _to_affine>](&mut affine.point, &point.point);
+                    }
+                    affine
+                }
+            }
+        }
+
+        impl $struct_name {
+            /// Return the distinguished generator point.
+            pub fn generator() -> Self {
+                unsafe {
+                    Self {
+                        point: *paste! { blst::[<blst_ $blst_name _generator>]() },
+                    }
+                }
+            }
+
+            pub fn from_raw(point: paste! { blst::[<blst_ $blst_name>] }) -> Self {
+                Self { point }
+            }
+
+            pub fn compress(&self) -> Vec<u8> {
+                let mut compressed_point = vec![0; $compressed_bytes];
+                let compress = paste! { blst::[<blst_ $blst_name _compress>] };
+                unsafe {
+                    compress(compressed_point.as_mut_ptr(), &self.point);
+                }
```

Comment:
any overhead here to defining the fn to a binding? rather than just assembling the name in place? there may be some limitations on `paste` here i don't know...

Reply:
I imagine the compiler optimises it, but there's no reason `paste!` couldn't be inside the `unsafe` block except that it breaks `rustfmt` so I was avoiding it for that reason.

There are probably a few `#[inline]` sprinklings and things that we can do to make sure we're squeezing every bit of performance out too